network/benchmark/src/lib.rs

```diff
@@ -0,0 +1,605 @@
+// Copyright © Aptos Foundation
+
+use aptos_config::{
+    config::NodeConfig,
+    network_id::{NetworkId, PeerNetworkId},
+};
+use aptos_logger::{info, warn};
+use aptos_metrics_core::{register_int_counter_vec, IntCounter, IntCounterVec};
+use aptos_network::{
+    application::interface::{NetworkClient, NetworkClientInterface, NetworkServiceEvents},
+    protocols::{network::Event, rpc::error::RpcError, wire::handshake::v1::ProtocolId},
+};
+use aptos_time_service::{TimeService, TimeServiceTrait};
+use aptos_types::{account_address::AccountAddress, PeerId};
+use bytes::Bytes;
+use futures::{
+    channel::oneshot::Sender,
+    stream::{FuturesUnordered, StreamExt},
+};
+use once_cell::sync::Lazy;
+use rand::{rngs::OsRng, Rng};
+use serde::{Deserialize, Serialize};
+use std::{ops::DerefMut, sync::Arc, time::Duration};
+use tokio::{runtime::Handle, select, sync::RwLock};
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[allow(clippy::large_enum_variant)]
+pub enum BenchmarkMessage {
+    DataSend(BenchmarkDataSend),
+    DataReply(BenchmarkDataReply),
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct BenchmarkDataSend {
+    pub request_counter: u64, // A monotonically increasing counter to verify responses
+    pub send_micros: i64, // micro seconds since some epoch at a moment just before this message is sent
+    pub data: Vec<u8>,    // A vector of bytes to send in the request; zero length in reply
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct BenchmarkDataReply {
+    pub request_counter: u64, // A monotonically increasing counter to verify responses
+    pub send_micros: i64, // micro seconds since some epoch at a moment just before this message is sent
+    pub your_send_micros: i64, // the send_micros from the previous message
```

Comment:
nit: maybe `request_send_micros`? I assume it's the timestamp for the corresponding request, right?